File: 378.kth-smallest-element-in-a-sorted-matrix.cpp
/*
 * @lc app=leetcode id=378 lang=cpp
 *
 * [378] Kth Smallest Element in a Sorted Matrix
 */

// @lc code=start
// class Solution {
// public:
//     using Point = pair<int, int>;
//     int kthSmallest(vector<vector<int>>& matrix, int k) {
//         int N = matrix.size();
//         auto cmp = [&matrix](Point& a, Point& b) {
//             return matrix[a.first][a.second] > matrix[b.first][b.second];
//         };
// 
//         priority_queue<Point, vector<Point>, decltype(cmp)> pq(cmp);
//         for(int j=0; j<N; ++j) {
//             pq.push({j, 0});
//         }
// 
//         while(k>1) {
//             auto curr = pq.top();
//             pq.pop();
// 
//             if(curr.second + 1 < N) {
//                 pq.push({curr.first, curr.second+1});
//             }
//             --k;
//         }
// 
//         auto ans = pq.top();
//         return matrix[ans.first][ans.second];
//     }
// 
// };
class Solution {
public:
    int kthSmallest(vector<vector<int>>& matrix, int k) {
        int N = matrix.size();
        int left = matrix[0][0], right = matrix[N-1][N-1];

        while(left < right) {
            int mid = left + (right - left) / 2;
            int less = matrix[0][0], more = matrix[N-1][N-1];
            int count = countLessEqual(matrix, mid, less, more);
            if(count == k) {
                return less;
            }
            if(count < k) {
                left = more;
            }
            else {
                right = less;
            }
        }
        return left;
    }

    int countLessEqual(vector<vector<int>>& matrix, int mid, int& less, int& more) {
        int N = matrix.size(), row = N-1, col = 0, count = 0;
        while(row>=0 and col<N) {
            if(matrix[row][col] > mid) {
                more = min(more, matrix[row][col]);
                --row;
            }
            else {
                less = max(less, matrix[row][col]);
                count += row + 1;
                ++col;
            }
        }
        return count;
    }
};
// @lc code=end

File: 378.kth-smallest-element-in-a-sorted-matrix.py
#
# @lc app=leetcode id=378 lang=python3
#
# [378] Kth Smallest Element in a Sorted Matrix
#

# @lc code=start
from typing import Tuple


class Solution:
    def kthSmallest(self, matrix: List[List[int]], k: int) -> int:
        N = len(matrix)
        minHeap = []

        for j in range(N):
            heapq.heappush(minHeap, (matrix[j][0], j, 0))
        
        while k:
            val, row, col = heapq.heappop(minHeap)
            if col+1 < N:
                heapq.heappush(minHeap, (matrix[row][col+1], row, col+1))
            k -= 1
        
        return val 


    # def kthSmallest(self, matrix: List[List[int]], k: int) -> int:
    #     self.matrix = matrix
    #     self.N = len(matrix)

    #     left = matrix[0][0]
    #     right = matrix[self.N-1][self.N-1]

    #     while left < right:
    #         mid = left + (right - left) // 2
    #         count, less, more = self.countLessEqual(mid)
    #         if count == k:
    #             return less
    #         elif count < k:
    #             left = more
    #         else:
    #             right = less

    #     return left

    #     
    # def countLessEqual(self, mid: int) -> Tuple[int, int, int]:
    #     row, col = self.N-1, 0
    #     more, less = self.matrix[self.N-1][self.N-1], self.matrix[0][0]
    #     count = 0
    #     while row >= 0 and col < self.N:
    #         if self.matrix[row][col] > mid:
    #             more = min(more, self.matrix[row][col])
    #             row -= 1
    #         else:
    #             less = max(less, self.matrix[row][col])
    #             count += row + 1
    #             col += 1
    #     
    #     return (count, less, more)



# @lc code=end

